[Mobile] - Refactor gallery - cherry pick v2 - step 1

mkevins opened the pull request:
### Related PRs

* Main refactor PR: https://github.com/WordPress/gutenberg/pull/25940
* Mobile refactor PR: https://github.com/WordPress/gutenberg/pull/31306
  * This PR is targeting the main refactor PR to bring the changes needed for the mobile implementations.
* Former mobile refactor PR: https://github.com/WordPress/gutenberg/pull/26823
  * Changes from this PR are being extracted and included in this and subsequent PRs

## Description
This PR is the first step at adapting the former refactor implementation to the new approach used in the main PR. It mostly cherry-picks changes to the rebased branch, and omits certain changes that were later reverted. After the inclusion of these changes, it should be possible to create new v2 gallery blocks (as inner image blocks), but the UI will not be in it's final state.

Also, a temporary change was included in this PR (https://github.com/WordPress/gutenberg/commit/acf18caab2f8e0788fe8d39d6d51d0751d7d50b6) to enable the `__experimentalGalleryRefactor` flag for testing purposes. This will be removed once a more appropriate implementation is in place.

The relevant commits from the former mobile refactor PR are included or omitted as follows:

Commit | Included?
-|-
WIP - https://github.com/WordPress/gutenberg/pull/26823/commits/57abff0071b3d1a5c843625bc13490156bd43954 | :heavy_check_mark:
fix dirty-image-options - https://github.com/WordPress/gutenberg/pull/26823/commits/795fd474be1d45bdbe3f2fb5c9db7ab49a6e0f40 | :x: These changes were later reverted for a change in approach, so they are not included here
fix spacing - https://github.com/WordPress/gutenberg/pull/26823/commits/2ec63f118da7312c547b7c3f4e2ed41ad4f7825c | :heavy_check_mark:
Add imageCrop, blockProps to block-list and style fixes - https://github.com/WordPress/gutenberg/pull/26823/commits/c0f043d49a9da0bdc79dfbeb0e4a639b515a2621 | :x: These changes were not included, but will be deferred to a later step.
Temporary remove the gallery from initial-html - https://github.com/WordPress/gutenberg/pull/26823/commits/c2fbfa86ea4d74871bac76d686f2a435553c23bb | :x: This was not included, since we probably want to opposite in order to test v1 galleries
Add gallery caption - https://github.com/WordPress/gutenberg/pull/26823/commits/051a2b832882ddb43fe71621d32663601fd30f87 | :heavy_check_mark:

## How has this been tested?
In the block editor, create a gallery and add some images. They should behave as inner image blocks, and should allow reordering and removal. It should also be possible to add captions to both the image blocks, and the outer gallery block. When an inner image block has a setting (attribute) that is also present on the gallery block, a change to the setting on the gallery level should "overwrite" the setting for the individual image block. E.g. if you add a link to an image, and subsequently change the gallery "link to" setting to "media file", the individual image link should no longer be set. However, if the same settings were applied first at the gallery level, then at the image level, the image block should retain its unique setting(s).

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

mkevins committed:
Make onFocus optional in MediaPlaceholder

mkevins committed:
Add useInnerBlocksProps hook

mkevins committed:
Enable __experimentalGalleryRefactor flag under __DEV__

This is only temporary, for testing purposes. This will be replaced with
an actual implementation (which will need to access the flag remotely).

mkevins committed:
Add WIP v2 gallery

mkevins committed:
Add numColumns to block-list flat list

mkevins committed:
Temporarily comment out spinner for pending imagesize option

This will need to be addressed (since we can't have unwrapped text on
mobile). For now, I'm commenting this out to proceed with cherry-picking
the changes from the original mobile refactor branch.

mkevins committed:
Fix spacing

mkevins committed:
Adjust styles to avoid appender overlap

mkevins committed:
Add gallery caption

antonis reviewed (approved):
Great work @mkevins 👍 
I tested this and the new gallery behaves pretty well considering the stage of this project:
* Created a gallery with some images
* Images behave as inner blocks
* Reordering works
* Removal works
* Attribute overwriting works as described 
* The columns setting seems to work similarly to the old v1 gallery block: Maximum number of columns is set to 3 and only two columns can be rendered in portrait mode (3 columns work only in landscape).

I also noticed the following:
* When an upload is cancelled, it will leave a placeholder image (as discussed). I don't think this is bad especially if we can add a way to resume the upload.
* I noticed a **breakage in the columns block** where the internal blocks are not outlined (screenshot) and the user is not able to add a block. I think we can tackle this in a separate PR.

<details>
 <summary>Columns block</summary>

![device-2021-05-03-194747](https://user-images.githubusercontent.com/304044/116906330-c9e84f80-ac48-11eb-89d4-266639932daf.png)
</details>

mkevins committed:
Fix lint

Some of these "fixes" are simply disabling lint for the offending lines.
These currently unused variables may be used in a later PR, so I'm
leaving them in, for now, to help simplify reconciling the changes from
the former refactor PR.

mkevins committed:
Fix sass lint

mkevins commented:
Thanks for reviewing and testing Antonis! Good catch about the columns. I will gather all of these issues as we progress further with the refactor PRs, and hopefully they'll get sorted out in further cherry-picking. It'll be great as a reference that you noted the issues here, since if the further cherry-picking from the former PR does _not_ resolve some issues, the comments on these PRs will serve as a good starting point to determine the root cause. In this case, I suspect it is to do with intermittent style changes to the block-list items, but I haven't confirmed that.

antonis commented:
> I suspect it is to do with intermittent style changes to the block-list items, but I haven't confirmed that.

That makes sense. I also tested the main refactor PR #25940 and the columns work there. Thus it must be some cherrypicked change on this one.
